arbitrator/prover/src/merkle.rs

```diff
@@ -43,81 +150,195 @@ impl MerkleType {
+            self.layers.lock().data[0].clone()
+        };
+        Merkle::new_advanced(self.ty, leaves, self.min_depth)
+    }
+}
+
 impl Merkle {
+    /// Creates a new Merkle tree with the given type and leaf hashes.
+    /// The tree is built up to the minimum depth necessary to hold all the
+    /// leaves.
     pub fn new(ty: MerkleType, hashes: Vec<Bytes32>) -> Merkle {
-        Self::new_advanced(ty, hashes, Bytes32::default(), 0)
+        Self::new_advanced(ty, hashes, 0)
     }
 
-    pub fn new_advanced(
-        ty: MerkleType,
-        hashes: Vec<Bytes32>,
-        empty_hash: Bytes32,
-        min_depth: usize,
-    ) -> Merkle {
-        if hashes.is_empty() {
+    /// Creates a new Merkle tree with the given type, leaf hashes, a hash to
+    /// use for representing empty leaves, and a minimum depth.
+    pub fn new_advanced(ty: MerkleType, hashes: Vec<Bytes32>, min_depth: usize) -> Merkle {
+        #[cfg(feature = "counters")]
+        NEW_COUNTERS[&ty].fetch_add(1, Ordering::Relaxed);
+        if hashes.is_empty() && min_depth == 0 {
             return Merkle::default();
         }
-        let mut layers = vec![hashes];
-        let mut empty_layers = vec![empty_hash];
+        let mut depth = (hashes.len() as f64).log2().ceil() as usize;
+        depth = depth.max(min_depth);
+        let mut layers: Vec<Vec<Bytes32>> = Vec::with_capacity(depth);
+        layers.push(hashes);
+        let mut dirty_indices: Vec<HashSet<usize>> = Vec::with_capacity(depth);
+        let mut layer_i = 0usize;
         while layers.last().unwrap().len() > 1 || layers.len() < min_depth {
-            let empty_layer = *empty_layers.last().unwrap();
-
-            #[cfg(feature = "rayon")]
-            let new_layer = layers.last().unwrap().par_chunks(2);
-
-            #[cfg(not(feature = "rayon"))]
-            let new_layer = layers.last().unwrap().chunks(2);
+            let layer = layers.last().unwrap();
+            let empty_hash = empty_hash_at(ty, layer_i);
 
-            let new_layer = new_layer
-                .map(|chunk| hash_node(ty, chunk[0], chunk.get(1).cloned().unwrap_or(empty_layer)))
-                .collect();
-            empty_layers.push(hash_node(ty, empty_layer, empty_layer));
+            let new_layer = new_layer(ty, layer, empty_hash);
+            dirty_indices.push(HashSet::with_capacity(new_layer.len()));
             layers.push(new_layer);
+            layer_i += 1;
         }
+        let layers = Mutex::new(Layers {
+            data: layers,
+            dirt: dirty_indices,
+        });
         Merkle {
             ty,
             layers,
-            empty_layers,
             min_depth,
         }
     }
 
+    fn rehash(&self, layers: &mut Layers) {
+        if layers.dirt.is_empty() || layers.dirt[0].is_empty() {
+            return;
+        }
+        for layer_i in 1..layers.data.len() {
+            let dirty_i = layer_i - 1;
+            let dirt = layers.dirt[dirty_i].clone();
```

Comment:
Cloning could be expensive here. Given you need to clear layers.dirt[dirty_i] near the end, consider `let mut dirt = std::mem::take(&mut layers.dirt[dirty_i]);` as that will give you the data without cloning and clear the original reference all in one go, that way, the call to `layers.dirt[dirty_i].clear();` at the end could be avoided

Reply:
Done